BrowserKit/Sources/TabDataStore/TabDataStore.swift

```diff
@@ -3,19 +3,208 @@
 // file, You can obtain one at http://mozilla.org/MPL/2.0/
 
 import Foundation
+import UIKit
+import Common
 
+//MARK: Protocol
 protocol TabDataStore {
-    func fetchTabData() async -> WindowData
-    func saveTabData(window: WindowData) async
-    func clearAllTabData() async
+    func fetchWindowData() async -> WindowData
+    func saveWindowData(window: WindowData) async
+    func clearAllWindowsData() async
 }
 
 actor DefaultTabDataStore: TabDataStore {
-    func fetchTabData() async -> WindowData {
+    //MARK: Variables
+    let browserKitInfo = BrowserKitInformation.shared
+    static let storePath = "codableWindowsState.archive"
+    private let saveQueue = DispatchQueue(label: "com.example.tabdatastore.save")
+    private var saveWorkItem: DispatchWorkItem?
+    private let throttleInterval: TimeInterval = 5
+    private var logger: Logger = DefaultLogger.shared
+
+    //MARK: URL Utils
+    private var windowDataDirectoryURL: URL? {
+        FileManager.default.containerURL(forSecurityApplicationGroupIdentifier: browserKitInfo.sharedContainerIdentifier)?
+            .appendingPathComponent("profile.profile")
+    }
+
+    private func windowURLPath(for windowId: UUID) -> URL? {
+        if let profileURL = FileManager.default.containerURL(forSecurityApplicationGroupIdentifier: browserKitInfo.sharedContainerIdentifier)?
+            .appendingPathComponent("profile.profile") {
+
+            do {
+                try FileManager.default.createDirectory(at: profileURL, withIntermediateDirectories: true, attributes: nil)
+            } catch {
+                logger.log("Error creating profile.profile directory: \(error)",
+                           level: .debug,
+                           category: .tabs)
+                return nil
+            }
+
+            let filePath = DefaultTabDataStore.storePath + "_\(windowId.uuidString)"
+            return profileURL.appendingPathComponent(filePath)
+        }
+        return nil
+    }
+
+    //MARK: Fetching Window Data
+    func fetchWindowData() async -> WindowData {
         return WindowData(id: UUID(), isPrimary: true, activeTabId: UUID(), tabData: [])
     }
 
-    func saveTabData(window: WindowData) async {}
+    func fetchWindowData(withId id: UUID) async -> WindowData? {
+        guard let profileURL = windowDataDirectoryURL else {
+            return nil
+        }
+
+        do {
+            let fileURLs = try FileManager.default.contentsOfDirectory(at: profileURL, includingPropertiesForKeys: nil, options: .skipsHiddenFiles)
+
+            let windowDataFiles = fileURLs.filter { $0.path.contains(DefaultTabDataStore.storePath) }
+
+            for fileURL in windowDataFiles {
+                do {
+                    let data = try Data(contentsOf: fileURL)
+                    let windowData = try JSONDecoder().decode(WindowData.self, from: data)
+                    if windowData.id == id {
+                        return windowData
+                    }
```

Comment:
I think there is some unnecessary work happening here, since the ID is in the url already we don't need to load the data or decode it to check it's id.